app/src/main/java/me/hufman/androidautoidrive/phoneui/SpotifyAuthorizationActivity.kt

```diff
@@ -0,0 +1,240 @@
+		Log.d(TAG, "Creating authorization service")
+		val builder = AppAuthConfiguration.Builder()
+		builder.setBrowserMatcher(BrowserWhitelist(
+				VersionedBrowserMatcher.CHROME_CUSTOM_TAB,
+				VersionedBrowserMatcher.SAMSUNG_CUSTOM_TAB))
+		builder.setConnectionBuilder(DefaultConnectionBuilder.INSTANCE)
+		return AuthorizationService(this, builder.build())
+	}
+
+	private fun processAuthorizationCode(requestCode: Int, resultCode: Int, data: Intent?) {
+		if (requestCode == REQUEST_CODE_SPOTIFY_LOGIN) {
+			if (resultCode == RESULT_CANCELED) {
+				onAuthorizationCancelled()
+				return
+			}
+
+			data?.let { intent ->
+				val response = AuthorizationResponse.fromIntent(intent)
+				val ex = AuthorizationException.fromIntent(intent)
+				when {
+					response?.authorizationCode != null -> {
+						authStateManager.updateAuthorizationResponse(response, ex)
+						onAuthorizationSucceed()
+					}
+					ex != null -> {
+						onAuthorizationFailed("Authorization flow failed: " + ex.message)
+					}
+					else -> {
+						onAuthorizationFailed("No authorization state retained - reauthorization required")
+					}
+				}
+			}
+		}
+	}
+
+	override fun onCreate(savedInstanceState: Bundle?) {
+		super.onCreate(savedInstanceState)
+		authStateManager = SpotifyAuthStateManager(this)
+		appSettingsReceiver = MutableAppSettingsReceiver(this)
+		initializeAppAuth()
+		doAuth()
+	}
+
+	override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+		super.onActivityResult(requestCode, resultCode, data)
+		processAuthorizationCode(requestCode, resultCode, data)
+	}
+
+	override fun onDestroy() {
+		super.onDestroy()
+		if (this::authService.isInitialized) {
+			authService.dispose()
+		}
+	}
+
+	/**
+	 * Authorized cancelled workflow. This is called if the authorization process activity was cancelled.
+	 */
+	private fun onAuthorizationCancelled() {
+		Log.d(TAG, "Authorization cancelled")
+		Toast.makeText(this, "Authorization canceled", Toast.LENGTH_SHORT).show()
+		finishActivityWithResult(AUTHORIZATION_CANCELED)
+	}
+
+	/**
+	 * Authorized failed workflow. This is called if the authorization process failed.
+	 */
+	private fun onAuthorizationFailed(error: String?) {
+		Log.d(TAG, "Authorization failed with the error: $error")
+		Toast.makeText(this, "Authorization failed", Toast.LENGTH_SHORT).show()
+		finishActivityWithResult(AUTHORIZATION_FAILED)
+	}
+
+	/**
+	 * Authorization success workflow. The [SpotifyWebApi] web API client instance will be initialized
+	 * with the authorized code obtained.
+	 */
+	private fun onAuthorizationSucceed() {
+		Log.d(TAG, "Authorization process completed successfully. AuthState updated")
+		clearNotAuthorizedNotification()
+		appSettingsReceiver[AppSettings.KEYS.SPOTIFY_SHOW_UNAUTHENTICATED_NOTIFICATION] = "true"
+		SpotifyWebApi.getInstance(this).initializeWebApi()
+		Toast.makeText(this, "Authorization successful", Toast.LENGTH_SHORT).show()
```

Comment:
I think `onAuthSucceeded` and `onAuthCanceled` should show a toast (non-intrusive because the user initiated it), while `onAuthError` should pop up the dialog.

Reply:
In any case, the toast messages should be translated, perhaps using the same strings that are present for the dialog in `strings.xml`

Reply:
Updated toast messages to use the translatable strings in strings.xml. I was thinking more about showing the error dialog on an authorization failure but the more I thought about it the more I realized that it didn't really go well with the flow. For example if the user is authorizing through clicking on the notification then opening a floating error dialog out of context is odd and seems out of place. In the case it fails I would imagine the user would go into the phone UI to see what happened and then pull up the error dialog (and also they will have to try again). Also when the authorization is started from the SpotifyApiErrorDialog itself using the button I don't think it would make sense to start another error dialog since that window will still be present and display the failure message above the button